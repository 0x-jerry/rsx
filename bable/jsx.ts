import BabelCore from '@babel/core'
import syntaxJsx from '@babel/plugin-syntax-jsx'
import t from '@babel/types'

const config = {
  excludePrefixReg: /^(on|_)/,
}

export function jsxPlugin({ types: t }: typeof BabelCore): BabelCore.PluginObj {
  return {
    name: 'babel-plugin-jsx',
    inherits: syntaxJsx.default,
    visitor: {
      JSXAttribute: {
        enter(path, state) {
          const valueNode = path.node.value
          const nameNode = path.node.name

          if (t.isJSXIdentifier(nameNode)) {
            const nodeName = nameNode.name

            if (config.excludePrefixReg.test(nodeName)) {
              return
            }
          }

          if (t.isJSXExpressionContainer(valueNode)) {
            convertJSXExpressionContainer(valueNode)
          }
        },
      },
      JSXElement: {
        enter(path) {
          for (const child of path.node.children) {
            if (t.isJSXExpressionContainer(child)) {
              convertJSXExpressionContainer(child)
            }
          }
        },
      },
    },
  }
}

function convertJSXExpressionContainer(node: t.JSXExpressionContainer) {
  if (t.isJSXEmptyExpression(node.expression)) {
    return
  }

  if (t.isLiteral(node.expression)) {
    return
  }

  const arrowFnExpress = t.arrowFunctionExpression([], node.expression)

  node.expression = arrowFnExpress

  return node
}
