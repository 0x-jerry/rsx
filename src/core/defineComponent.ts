import { isFn } from '@0x-jerry/utils'
import type { Merge } from 'type-fest'
import type { ToMaybeRef } from '.'
import type { DefineProps } from './props'

export interface PropOption<T = any> {
  type?: T
  default?: T | (() => T)
}

export type PropsType<T = any> = Record<string, T | PropOption<T>>

export interface ComponentOption<Props extends PropsType = {}> {
  props?: Props
  toRefs?: boolean
}

interface CommonProps {
  class?: any
  [key: string]: any
}

export type FunctionalComponent<P extends PropsType = {}> = (
  props: Merge<P, CommonProps>,
  children?: any[],
) => JSX.Element

/**
 * without option
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * with option
 * @param opt
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P>,
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * implement
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P> | FunctionalComponent,
  impl?: FunctionalComponent,
): FunctionalComponent {
  if (isFn(opt)) return opt

  // todo check props type
  return impl!
}

export const dc = defineComponent
