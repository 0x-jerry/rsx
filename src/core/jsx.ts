import { camelCase, isString, PascalCase } from '@0x-jerry/utils'
import { isRef, unref } from '@vue/reactivity'
import {
  appendToCurrentContext,
  createNodeContext,
  popCurrentContext,
  setCurrentContext,
} from './context'
import {
  createFragment,
  createNativeElement,
  type DComponent,
  type DNode,
  isDComponent,
} from './node'

type FunctionalComponent = (props?: any, children?: DNode[]) => DComponent

export function h(
  type: string | FunctionalComponent | DComponent,
  props?: Record<string, any>,
  ...children: DNode[]
): DComponent {
  if (isDComponent(type)) {
    return type
  }

  const _props = transformProps(type, props)

  if (type === Fragment) {
    return type(_props, children)
  }

  if (isString(type)) {
    return createNativeElement(type, _props, children)
  }

  return createComponentInstance(type, _props, children)
}

export const Fragment: FunctionalComponent = (_, children) => {
  return createFragment(children || [])
}

function createComponentInstance(
  type: FunctionalComponent,
  props?: any,
  children?: DNode[],
) {
  const ctx = createNodeContext(type.name)

  appendToCurrentContext(ctx)

  setCurrentContext(ctx)

  const el = type(props, children)

  el._ = ctx

  popCurrentContext()
  return el
}

function transformProps(type: any, props?: Record<string, any>): any {
  const _raw: Record<string, any> = {}

  if (!props) return _raw

  Object.entries(props).forEach(([key, value]) => {
    if (!key.startsWith('$')) {
      // Prevent change props directly.
      _raw[key] = value
      return
    }

    if (key === '$') {
      const newProps = transformDefaultBinding(type, value, props)

      // todo, check duplicate key
      Object.assign(_raw, newProps)
      return
    }

    // normal binding syntax sugar
    // <input $value:trim={refValue} />

    const [name, modifier] = key.slice(1).split(':')
    _raw[camelCase(name)] = value

    if (isRef(value)) {
      // fix me: compose events
      _raw[`onUpdate${PascalCase(name)}`] = (v: unknown) => (value.value = v)
    }
  })

  // return _raw
  return new Proxy(_raw, {
    get(t, p, r) {
      return unref(Reflect.get(t, p, r))
    },
    set() {
      // Prevent change props directly.
      return true
    },
  })
}

// todo, default binding syntax sugar
// current: $xx={refValue}
// suppport new sugar: $xx={[data, 'key']}
function transformDefaultBinding(
  type: any,
  value: any,
  allProps: Record<string, any>,
) {
  const props: Record<string, any> = {}

  if (type === 'input') {
    if (unref(allProps.type) === 'checkbox') {
      props.checked = value
      if (isRef(value)) {
        props.onChange = (e: InputEvent) => {
          value.value = (e.target as HTMLInputElement).checked
          console.log('checked', value.value)
        }
      }
    } else {
      props.value = value

      if (isRef(value)) {
        props.onInput = (e: InputEvent) =>
          (value.value = (e.target as HTMLInputElement).value)
      }
    }
  } else if (type === 'select') {
    props.value = value

    if (isRef(value)) {
      props.onChange = (e: InputEvent) =>
        (value.value = (e.target as HTMLSelectElement).value)
    }
  }

  return props
}
