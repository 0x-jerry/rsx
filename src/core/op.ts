import { isString } from '@0x-jerry/utils'
import { disableSSR, enableSSR } from './config'
import {
  appendToCurrentContext,
  createNodeContext,
  popCurrentContext,
  setCurrentContext,
} from './context'
import type { FunctionalComponent } from './defineComponent'
import { mount } from './hook'
import { moveTo } from './nodeOp'

export function mountApp(
  App: FunctionalComponent,
  selector: string | HTMLElement,
) {
  const container = isString(selector)
    ? document.querySelector(selector)
    : selector

  if (!container) {
    console.warn(`Can't find container`)
    return
  }

  const ctx = createNodeContext('Root')

  appendToCurrentContext(ctx)

  setCurrentContext(ctx)

  const el = App({}, [])

  ctx.el = el

  moveTo(container, el)

  popCurrentContext()

  mount(ctx)
}

export function renderToString(Comp: FunctionalComponent<any>) {
  // todo
  enableSSR()

  const app = document.createElement('div')
  mountApp(Comp, app)

  disableSSR()

  return app.innerHTML
}
