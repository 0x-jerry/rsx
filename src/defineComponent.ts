import type { EmptyObject } from '@0x-jerry/utils'
import type { DefineProps } from './props'
import type { IsStartWithCapitalizedLetter, MaybeRef } from './types'

export type PropsType<T = any> = Record<string, T>

type ToPropRefs<T extends {}> = {
  /**
   * Start with capitalized letter means this prop is not used as reactive prop
   */
  [key in keyof T]: key extends IsStartWithCapitalizedLetter<key>
    ? T[key]
    : MaybeRef<T[key]>
}

/**
 * @private
 */
export type FunctionalComponent<P extends PropsType = any> = (
  props: P,
  children?: any[],
) => unknown

type _NoInferProps<P extends EmptyObject> = DefineProps<P> & Record<string, any>

export type ExposedFunctionalComponent<P extends PropsType = any> =
  FunctionalComponent<ToPropRefs<P>>

/**
 * Auto calculate props from generic type P
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: NoInfer<FunctionalComponent<_NoInferProps<P>>>,
): ExposedFunctionalComponent<P>

/**
 * Infer props from impl parameter
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<P>,
): ExposedFunctionalComponent<P>

/**
 * implement
 */
export function defineComponent(impl?: any): any {
  // todo check props type
  return impl
}

export const dc = defineComponent
