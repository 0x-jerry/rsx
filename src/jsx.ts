import { isString } from '@0x-jerry/utils'
import type { FunctionalComponent } from './defineComponent'
import { Fragment } from './internalComponents'
import { createFragment } from './nodes/AnchorNode'
import { createComponentNode } from './nodes/ComponentNode'
import { createNativeNode } from './nodes/NativeNode'
import type { NodeElement } from './nodes/shared'
import type { AnyProps } from './props'

export function h(
  type: string | FunctionalComponent,
  props?: AnyProps,
  ...children: unknown[]
): NodeElement {
  if (isString(type)) {
    return createNativeNode(type, props, children)
  }

  if (type === Fragment) {
    return createFragment(children)
  }

  return createComponentNode(type, props, children)
}
