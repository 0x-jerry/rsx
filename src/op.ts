import { isString } from '@0x-jerry/utils'
import { disableSSR, enableSSR } from './config'
import type { FunctionalComponent } from './defineComponent'
import { mount, unmount } from './hook'
import { h } from './jsx'
import { moveTo } from './nodeOp'
import type { ComponentNode } from './nodes/ComponentNode'

export function mountApp(
  App: FunctionalComponent,
  selector: string | HTMLElement,
) {
  const container = isString(selector)
    ? document.querySelector(selector)
    : selector

  if (!container) {
    throw new Error(`Can't find container`)
  }

  const rootEl = h(App) as ComponentNode

  rootEl.initialize()

  if (rootEl.el) {
    moveTo(container, rootEl.el)
  }

  mount(rootEl.context)

  return rootEl
}

export function unmountApp(app: ComponentNode) {
  if (app.context._unmounted) {
    return
  }

  unmount(app.context)
}

export function renderToString(Comp: FunctionalComponent) {
  // todo
  enableSSR()

  const app = document.createElement('div')
  mountApp(Comp, app)

  disableSSR()

  return app.innerHTML
}
