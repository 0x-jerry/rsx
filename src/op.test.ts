import { dc } from './defineComponent'
import { h } from './jsx'
import { renderToString } from './op'

describe('ssr', () => {
  it('should render as string', async () => {
    const Comp = dc(() => {
      return h('div', { class: 'test' }, 'hello world!')
    })

    const str = renderToString(Comp)
    expect(str).toMatchSnapshot()
  })
})
