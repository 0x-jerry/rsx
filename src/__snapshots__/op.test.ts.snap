// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ssr > should render as string > html 1`] = `"<div class="test">hello world!</div>"`;
